### PR TITLE
MerkleDB documentation

### PR DESCRIPTION
**Pull request by Maria-Nosyk**

### CODE REVIEW

**Comment by aleksuss** [comment]
@Maria-Nosyk fix a typo and formatting please

File: src/get-started/design-overview.md [hunk 1]
```diff
@@ -45,9 +45,10 @@ are applied to the key-value storage.
 
 All data in the Exonum blockchain is divided into two parts:
 
-- **Data storage**, which contains data structured into tables
+- **Data storage**, which contains data structured into objects. Objects
+  represent high-level wrappers over the key key-value store
```
Comment (dmitry-timofeev) [inline src/get-started/design-overview.md]: ~~key~~ key-value store

File: src/get-started/design-overview.md [hunk 2]
```diff
@@ -224,62 +225,67 @@ which more than 1/3 (but less than 2/3) of the validators are compromised.
...
-adding and removing elements from the set. Finally, all collections support
-iterations over items (or keys, values, and key-value pairs in the case of
-  maps).
+  Both `ListIndex` and `ProofListIndex` support updating by index and
+  appending only. `MapIndex` and `ProofMapIndex` allow inserting,
+  updating or deleting key-value pairs. `KeySetIndex` and `ValueSetIndex`
+  support adding and removing elements from the set. Finally, all collections
+  support iterations over items (or keys, values, and key-value pairs in the
+  case of maps).
+- `Entry` represents a set that contains only one item.
```
Comment (dmitry-timofeev) [inline src/get-started/design-overview.md]: Is it a singleton set, or is it akin to an `Optional` which can contain either one or zero values?

File: src/get-started/design-overview.md [hunk 3]
```diff
@@ -224,62 +225,67 @@ which more than 1/3 (but less than 2/3) of the validators are compromised.
...
   `ValueSetIndex` is a better match for complex items or items that need to be
   looked up by a hash.
 
-Both `ListIndex` and `ProofListIndex` support updating by index and
-appending only. `MapIndex` and `ProofMapIndex` allow inserting,
-updating or deleting key-value pairs. `KeySetIndex` and `ValueSetIndex` support
-adding and removing elements from the set. Finally, all collections support
-iterations over items (or keys, values, and key-value pairs in the case of
-  maps).
+  Both `ListIndex` and `ProofListIndex` support updating by index and
```
Comment (dmitry-timofeev) [inline src/get-started/design-overview.md]: ListIndex used to support removing any number of elements from the tail. Is this no longer the case?

File: src/get-started/design-overview.md [hunk 4]
```diff
@@ -224,62 +225,67 @@ which more than 1/3 (but less than 2/3) of the validators are compromised.
...
+- `Entry` represents a set that contains only one item.
 
 ### Proofs
 
 `ProofListIndex` and `ProofMapIndex` allow efficiently
 creating a proof that specific values are saved under particular keys.
 To prove that, it is sufficient to return a list of hashes from
 the tree root to a particular cell (a Merkle path). Merkle Patricia
-tables also allow to generate proofs that there is no data in the
+stores also allow to generate proofs that there is no data in the
```
Comment (dmitry-timofeev) [inline src/get-started/design-overview.md]: IMO, that sounds somewhat strange — like a place with a fancy name (do we use a 'store' elsewhere?)

File: src/architecture/storage.md [hunk 5]
```diff
@@ -1,262 +1,158 @@
...
-The following additional procedures are implemented:
+Currently [RocksDB][rocks-db] is used for low-level key-value storage engine in
+Exonum.
 
-- Get the height of the Merkle tree. As the tree is balanced (though may be not
-  full), its height is close to `log2` of the list length
-- Get the value of the tree root (i.e., the hash of the entire Merkle tree)
-- Build a proof of existence for an item at a specific position
-- Build a proof of existence for items at a specific contiguous index range
+The objects of MerkleDB are convenient abstractions for work with
```
Comment (dmitry-timofeev) [inline src/architecture/storage.md]: > for work with blockchain-specific data

Doesn't sound right.

File: src/architecture/storage.md [hunk 6]
```diff
@@ -1,262 +1,158 @@
...
-    `ProofListIndex` also persists all intermediate nodes of the Merkle tree
-    built on top of the list, in order to quickly build proofs and recalculate
-    the Merkle tree after operations on the list.
+- **blobs**, which represent sequences of bytes, and
+- **root objects**, that do not have parents. These objects have UTF-8
+  identifiers, for example, "block", "state". Root objects can contain blob
+  items inside them.
 
-#### ProofMapIndex
+The basic root objects of the framework are:
```
Comment (dmitry-timofeev) [inline src/architecture/storage.md]: Is it really needed to use 'root object' at this point of the db development? Will non-root object be of different types (**not** lists, maps, …)? If there is nothing else, shall a user care if they are 'root'? 

Do I also have multiple roots — a forest of these objects?

File: src/architecture/storage.md [hunk 7]
```diff
@@ -1,262 +1,158 @@
...
-Currently the main database engine is [RocksDB][rocks-db].
+- `transactions: MapIndex`  
+  Represents a map from the transaction hash into a raw transaction structure.
+- `transaction_results: ProofMapIndex`  
+  Keeps execution results for all accepted transactions,
+  indexed by transaction hashes.
+- `transactions_pool: KeySetIndex`  
+  Stores the set of hashes of the known transactions that have not been
+  committed yet.
+- `transactions_pool_len: Entry`
```
Comment (dmitry-timofeev) [inline src/architecture/storage.md]: :fearful: Can't the keyset do it on its own?

File: src/architecture/storage.md [hunk 8]
```diff
@@ -1,262 +1,158 @@
...
+  block and the block height.
+- `blocks: MapIndex`  
+  Stores the block object for every block height.
+- `block_hashes_by_height: ListIndex`  
+  Saves block hashes indexed by block heights.
+- `block_transactions: ProofListIndex`  
+  Stores a list of transactions of the specific block indexed by the block
+  height.
+- `precommits: ListIndex`  
+  Stores a list of precommits of the validators for the specific block, indexed
```
Comment (dmitry-timofeev) [inline src/architecture/storage.md]: > indexed by the block hash.

That sounds confusing without any introduction to families (is that still a thing?). How a _list_ is indexed by a _hash_?

File: src/architecture/storage.md [hunk 9]
```diff
@@ -1,262 +1,158 @@
...
-Every table is uniquely identified by a compound identifier, which is used
-to map table keys into a column family and its keys in the underlying
-low-level storage. A table identifier consists of 2 parts:
+To use a particular database, a minimal [`Database`][database]
+interface should be implemented for it:
 
-- **String name,** which is mapped 1-to-1 to a column family.
-  The name may contain uppercase and lowercase Latin letters, digits,
-  underscores `_`, and periods `.`. By convention, table names in services
+- Get a value by a [column family][col-family] name and a key
```
Comment (dmitry-timofeev) [inline src/architecture/storage.md]: Is it appropriate to link Rocksdb-specific article when talking about _our_ interface for plugging any database engines?

File: src/architecture/storage.md [hunk 10]
```diff
@@ -1,262 +1,158 @@
...
+- Put a new value at the specified column family / key (insert or update
+  the saved one)
+- Delete a key-value pair by a column family name / key.
+
+All the objects functionality is reduced to these atomic call types.
+
+Values of different objects and their items are stored in a single column
+family in the
+low-level storage. Keys are represented as byte sequences. Values are
+serialized according to the Protobuf serialization format. All keys are mapped
```
Comment (dmitry-timofeev) [inline src/architecture/storage.md]: Is it appropriate (at this level) to say that it enforces a certain serialization format? Isn't that/Shan't that be provided by the upper levels (i.e., maps/sets/etc)?

File: src/architecture/storage.md [hunk 11]
```diff
@@ -1,262 +1,158 @@
...
+  underscores `_`, and periods `.`. By convention, object names in services
   should
   start with [the service name][service-name] and a period. For example,
-  the only table in the Cryptocurrency Tutorial is named
+  the only object in the Cryptocurrency Tutorial is named
   `cryptocurrency.wallets`,
   where `cryptocurrency` is the service name, and `wallets` is the own name
-  of the table.
+  of the object.
 - **Optional prefix** presented as a sequence of bytes (`Vec<u8>` in Rust
```
Comment (dmitry-timofeev) [inline src/architecture/storage.md]: This documentation doesn't seem to be detailed enough — above we present half a dozen of system collections that use prefixes, but don't explain what it is.

File: src/architecture/storage.md [hunk 12]
```diff
@@ -1,262 +1,158 @@
...
-    `exchange.crypto`.
-
-!!! warning
-    It is strongly advised not to admit
-    a situation when a table prefix in a table group starts with
-    another table prefix in the same group. Such cases may cause unpredictable
-    collisions between logically different keys and elements.
-    As a possible way to avoid this, prefixes within the group may have
-    a fixed byte size.
+    Suppose we have a list with the address `(" exchange.crypto "," BTC ")` in
```
Comment (dmitry-timofeev) [inline src/architecture/storage.md]: The paragraph above says that 

> Key in this column family [object pool] is the object identifier

Then how are names mapped to object identifiers?

File: src/architecture/storage.md [hunk 13]
```diff
@@ -269,7 +165,7 @@ operations: put a value addressed by a key, or delete a value by a key.
 
 [Snapshot][snapshot] fixes the storage state at the moment of snapshot
 creation and provides a read-only API to it. Even if the storage state
-is updated, the snapshot still refers to the old table content.
+is updated, the snapshot still refers to the old object content.
```
Comment (dmitry-timofeev) [inline src/architecture/storage.md]: Which object? Any number of objects can be accessed with the snapshot?

File: misc/cspell.json [hunk 14]
```diff
@@ -39,6 +39,7 @@
     "healthcheck",
     "HMACs",
     "idempotence",
+	   "indexes",
```
Comment (slowli) [inline misc/cspell.json]: Nit: extra space at the beginning of the line.

File: src/architecture/merkledb.md [hunk 15]
```diff
@@ -0,0 +1,207 @@
...
+for the [data storage](storage.md). All objects fall into the following groups:
+
+- **blobs** which represent sequences of bytes
+- **root objects** that do not have parents. These objects have UTF-8
+  identifiers, for example, "block", "state". Root objects can contain blob
+  items inside them.
+
+## Root Object Types
+
+Root objects perform the same role as in relational database
```
Comment (dmitry-timofeev) [inline src/architecture/merkledb.md]: What role? This paragraph doesn't say that, and I'm not sure if this comparison helps without that.

File: src/architecture/merkledb.md [hunk 16]
```diff
@@ -0,0 +1,207 @@
...
+management systems (RDBMSs). However, unlike RDBMS tables,
+all MerkleDB objects internally are implemented as wrappers around key-value
```
Comment (dmitry-timofeev) [inline src/architecture/merkledb.md]: Shall we reveal the implementation details in the introduction (or at all)?

File: src/architecture/merkledb.md [hunk 17]
```diff
@@ -0,0 +1,207 @@
...
+- **map**, where key-value pairs are stored
+- **set** of unique items
+- **entry**, which represents a set containing only one item.
+
+Both keys and values in the wrapped stores are persisted as byte sequences.
+MerkleDB does not natively support operations (matching, grouping, sorting,
+etc.) over separate value fields, as it is the case with other key-value
+storages.
+
+### Key Sorting and Iterators
```
Comment (dmitry-timofeev) [inline src/architecture/merkledb.md]: I am not sure this subsection works here. First, it is too abstract — it requires three pieces of knowledge to understand the iteration order — what is used as keys (not that clear for lists, and sets), how they are serialized, and what is lexicographical order of that. Isn't that too much to ask from a new user? Second — is it important at this point? Shan't it be documented in the iterators? For example, in KeySetIndex#iter, that the elements are ordered by the lexicographical order of their (SHA-256) hashes.

File: src/architecture/merkledb.md [hunk 18]
```diff
@@ -0,0 +1,207 @@
...
+
+### MapIndex
+
+[`MapIndex`][map-index] implements a key-value store aka a map. It has
+the following functionality:
+
+- Get, set and remove value by key
+- Check if a specific key is present in the map
+- Iterate over the key-value pairs in the lexicographic key order
+- Iterate over keys in the lexicographic key order
```
Comment (dmitry-timofeev) [inline src/architecture/merkledb.md]: What if we join these together:

"Iterate over keys, values and entries in …"

File: src/architecture/merkledb.md [hunk 19]
```diff
@@ -0,0 +1,207 @@
...
+[`ListIndex`][list-index] represents an array list.
+The following operations are supported:
+
+- Get and set a list item by index
+- Append an item to the list
+- Pop or poll the last item from the list
+- Get the list length
+- Check if the list is empty
+- Iterate over index-item pairs ordered by index
+- Insert a sequence of items from an iterator
```
Comment (dmitry-timofeev) [inline src/architecture/merkledb.md]: We usually insert at a particular index, and append/add/push to the end of the list. If I understand correctly, the former is not supported at all.

File: src/architecture/merkledb.md [hunk 20]
```diff
@@ -0,0 +1,207 @@
...
+    and items themselves as corresponding values.
+
+### KeySetIndex
+
+[`KeySetIndex`][key-set-index] represents a set.
+The following procedures are implemented:
+
+- Add and remove set items
+- Check if a specific item is in the set
+- Iterate over items in the lexicographic order
```
Comment (dmitry-timofeev) [inline src/architecture/merkledb.md]: Shall we clarify that 

"in the lexicographic order of their serialized/binary representation"?

File: src/architecture/merkledb.md [hunk 21]
```diff
@@ -0,0 +1,207 @@
...
+  small
+  tuples). On the other hand, `ValueSetIndex` handles storing big and complex
+  items more easily.
+- `KeySetIndex` introduces a lexicographical order over stored
+  items. In `ValueSetIndex` items are ordered according to their hash
+  function properties.
+
+### Entry
+
+`Entry` represents a set that contains only one item.
```
Comment (dmitry-timofeev) [inline src/architecture/merkledb.md]: I wouldn't call it a set as it is more akin to an `Optional` — a container that may contain a single item or be empty (= no item).

File: src/architecture/merkledb.md [hunk 22]
```diff
@@ -0,0 +1,207 @@
...
+built and sent along with the actual data. Having block headers
+and this proof, the client may check that the received data was really
+authorized by the validators without having to replicate the entire blockchain
+contents.
+
+#### ProofListIndex
+
+[`ProofListIndex`][proof-list-index] implements a Merkle
+tree, which is a merkelized version of an
+array list. It implements the same methods as `ListIndex`, and adds an
```
Comment (dmitry-timofeev) [inline src/architecture/merkledb.md]: This is not entirely true, according to the note below (ListIndex supports removal from the tail):

> `ProofListIndex` is an *append-only* store; it does not allow deleting list items.

File: src/architecture/merkledb.md [hunk 23]
```diff
@@ -0,0 +1,207 @@
...
+additional feature: based on Merkle trees, `ProofListIndex` allows efficiently
+creating compact proofs of existence for the list items.
+The following additional procedures are implemented:
+
+- Get the height of the Merkle tree. As the tree is balanced (though may be not
+  full), its height is close to `log2` of the list length
+- Get the value of the tree root (i.e., the hash of the entire Merkle tree)
+- Build a proof of existence/absence for an item at a specific position
```
Comment (dmitry-timofeev) [inline src/architecture/merkledb.md]: If I remember correctly, out-of-range proofs couldn't be built. Has this been implemented?

File: src/architecture/storage.md (hunk 8 above)
Comment (Maria-Nosyk) [reply src/architecture/storage.md]: edited

File: src/architecture/storage.md (hunk 10 above)
Comment (Maria-Nosyk) [reply src/architecture/storage.md]: moved

File: src/architecture/merkledb.md (hunk 23 above)
Comment (pavel-mukhanov) [reply src/architecture/merkledb.md]: Yes, we implemented it, you can now get proof of absence of a single element or range from `ProofListIndex`.

File: src/architecture/storage.md (hunk 6 above)
Comment (Maria-Nosyk) [reply src/architecture/storage.md]: Here we need to explicitly distinguish between blobs and other objects. Besides, in the future, if I understand correctly, the term "root object" will also distinguish the object that has no parents from other objects.

There can be multiple root objects within one service.

File: src/architecture/merkledb.md (hunk 23 above)
Comment (dmitry-timofeev) [reply src/architecture/merkledb.md]: That's neat :+1:

File: src/architecture/storage.md [hunk 24]
```diff
@@ -1,262 +1,158 @@
...
+  of the object.
 - **Optional prefix** presented as a sequence of bytes (`Vec<u8>` in Rust
   terms).
 
-If the prefix is present, the column family identified by the table name
-stores a *group* of tables, rather than a single table.
-In this case, prefixes are used to distinguish tables within the group.
+An object key in RocksDB is a 32-elements byte array. The first 16 bytes are an
+object identifier
+obtained using the object pool. The second 16 bytes are the key of an item
```
[the remark this answers is not in the document]
Comment (Maria-Nosyk) [reply src/architecture/storage.md]: done

File: src/architecture/storage.md (hunk 12 above)
Comment (Maria-Nosyk) [reply src/architecture/storage.md]: edited

File: src/architecture/storage.md [hunk 25]
```diff
@@ -1,262 +1,158 @@
...
-    collisions between logically different keys and elements.
-    As a possible way to avoid this, prefixes within the group may have
-    a fixed byte size.
+    Suppose we have a list with the address `(" exchange.crypto "," BTC ")` in
+    which we put one value, for example `0.00019`. The pool assigns
+    0x0000000000000003 identifier to this list, when we create it.
+    Since `0.00019` is
+    a single value in the list, its key is 0x0000000000000000. Thus, in the
+    database it will look like this (in HEX):
+    0x00000000000000030000000000000000 : 0x302E3030303139.
```
[the remark this answers is not in the document]
Comment (Maria-Nosyk) [reply src/architecture/storage.md]: example changed

File: src/architecture/storage.md (hunk 7 above)
Comment (Maria-Nosyk) [reply src/architecture/storage.md]: deleted this object from the list

File: src/architecture/merkledb.md (hunk 18 above)
Comment (Maria-Nosyk) [reply src/architecture/merkledb.md]: > Also now there is possible to get total number of items for every collection type.

done

File: src/architecture/merkledb.md (hunk 19 above)
Comment (Maria-Nosyk) [reply src/architecture/merkledb.md]: corrected

File: src/architecture/merkledb.md (hunk 17 above)
Comment (Maria-Nosyk) [reply src/architecture/merkledb.md]: the section moved where more appropriate

File: src/architecture/storage.md [hunk 26]
```diff
@@ -1,262 +1,158 @@
...
   where `cryptocurrency` is the service name, and `wallets` is the own name
-  of the table.
+  of the object.
 - **Optional prefix** presented as a sequence of bytes (`Vec<u8>` in Rust
   terms).
 
-If the prefix is present, the column family identified by the table name
-stores a *group* of tables, rather than a single table.
-In this case, prefixes are used to distinguish tables within the group.
+An object key in RocksDB is a 32-elements byte array. The first 16 bytes are an
```
[the remark this answers is not in the document]
Comment (Maria-Nosyk) [reply src/architecture/storage.md]: example changed. information on the number of bytes was eliminated

File: src/architecture/merkledb.md [hunk 27]
```diff
@@ -0,0 +1,348 @@
...
+    Suppose we have a list with the address `(" exchange.crypto "," BTC ")` in
+    which we put one value, for example `7865`. The pool assigns identifier
+    `3` to this list when we create it. In the database the identifier looks
+    like this - `0x0000000000000003`.
+
+    Since `7865` is a single value in the created list, its index in the
+    database looks like this - `0x0000000000000000`.
+
+    Thus, the whole key in the database will look as follows (in HEX) -
+    `0x00000000000000030000000000000000 : 0x302E3030303139`.
```
Comment (slowli) [inline src/architecture/merkledb.md]: > `0x302E3030303139`

? This is ASCII / UTF-8 encoding of a string `0.00019`. How does it correspond to 7865? As an example, 7865 as a 64-bit signed (or unsigned) little-endian integer is encoded as `0x_bd1e_0000_0000_0000`.

File: src/architecture/merkledb.md [hunk 28]
```diff
@@ -0,0 +1,348 @@
...
+
+- an **object identifier** obtained from the object pool
+- an **index of an item** inside the object.
+
+Both parts of the key are encoded as big-endian.
+
+!!! note "Example"
+    Suppose we have a list with the address `(" exchange.crypto "," BTC ")` in
+    which we put one value, for example `7865`. The pool assigns identifier
+    `3` to this list when we create it. In the database the identifier looks
```
Comment (slowli) [inline src/architecture/merkledb.md]: Nit: I'd be more clear that the assigned identifier is pseudo-random; a reader may get the idea that `3` is derived from the address.

File: src/architecture/storage.md [hunk 29]
```diff
@@ -1,301 +1,40 @@
...
-   for data storage
-2. [Low-level storage](#low-level-storage) explains how tables are persisted
-   using RocksDB
-3. [View layer](#view-layer) describes the wrapper over the DB engine
-   that ensures atomicity of blocks and transactions
-4. [List of system tables](#system-tables) contains tables
-   used directly by the Exonum core
-5. [Indexing](#indexing) gives an insight how indices over structured data
-   can be built in Exonum
+Exonum uses MerkleDB as a storage framework. MerkleDB is an object database. It
```
Comment (slowli) [inline src/architecture/storage.md]: Nit: Link to the MerkleDB page?

File: src/architecture/storage.md [hunk 30]
```diff
@@ -1,301 +1,40 @@
...
+represents a high-level wrapper over the key-value store.
 
-## Table Types
+Currently [RocksDB][rocks-db] is used for low-level key-value storage engine in
```
Comment (slowli) [inline src/architecture/storage.md]: Nit: I'd move this to the previous statement:

> [...] a high-level wrapper over the key-value store (RocksDB).

The present intro formatting seems too fragmented to me.

File: src/architecture/storage.md [hunk 31]
```diff
@@ -1,301 +1,40 @@
...
-Exonum tables implement iterators over stored items (or keys, values, and
-key-value
-pairs in the case of maps). Such
-iterators use key ordering of the underlying key-value storage to determine
-the iteration order.
-Namely, keys are lexicographically ordered over their binary serializations;
-this ordering coincides with that used in
-[RocksDB](#low-level-storage).
+- **blobs**, which represent sequences of bytes, and
+- **root objects**, that do not have parents. These objects have UTF-8
```
Comment (slowli) [inline src/architecture/storage.md]: Nit: shouldn't it be "UTF-8 string identifiers"?

File: src/architecture/merkledb.md [hunk 32]
```diff
@@ -0,0 +1,348 @@
...
+
+MerkleDB is an object database. Objects represent the highest abstraction level
+for the [data storage](storage.md). All objects fall into the following groups:
+
+- **blobs** which represent sequences of bytes
+- **root objects** that do not have parents. These objects have UTF-8
+  identifiers, for example, "block", "state". Root objects can contain blob
+  items inside them.
+
+1. [Root Objects Types](#root-object-types) lists the types of data
```
Comment (slowli) [inline src/architecture/merkledb.md]: Weird enough, my local copy of `mkdocs` glues this list with a previous one, resulting in quite incomprehensible formatting. I'd suggest to insert a paragraph before this list (something like "The rest of the article is structured as follows") to get rid of this bug (?) and make the formatting more appealing in general.

File: src/architecture/merkledb.md [hunk 33]
```diff
@@ -0,0 +1,348 @@
...
+
+## Root Object Types
+
+Root objects perform the same role as tables in relational database
+management systems (RDBMSs). All MerkleDB objects internally are
+implemented as wrappers around key-value stores.
+
+Root objects in MerkleDB fall into several types:
+
+- **list** of items. Each item is preceded by an index which represents a `u64`
```
Comment (slowli) [inline src/architecture/merkledb.md]: > Each item is preceded by an index

I'd omit such details; IMO, they create more confusion than clarify things (e.g., it's not clear *where* the index precedes an item).

File: src/architecture/merkledb.md [hunk 34]
```diff
@@ -0,0 +1,348 @@
...
+  integer. Indices are implicitly defined by the items order
+- **map**, where key-value pairs are stored
+- **set** of unique items
+- **entry**, which represents an optional single item.
+
+Both keys and values in the wrapped stores are persisted as byte sequences.
+Values are serialized according to the Protobuf serialization format.
```
Comment (slowli) [inline src/architecture/merkledb.md]: Nit: this is opt-in and (for better or worse) is not enforced on the MerkleDB level. For example, some core objects store hashes as values, and they are not stored as Protobuf's `bytes` or messages with a single `bytes` field, but rather directly as byte sequences.

File: src/architecture/merkledb.md [hunk 35]
```diff
@@ -0,0 +1,348 @@
...
+the outer point of view, the changes are eagerly applied to the data
+storage; however, these changes are stored directly in the fork and may
+be easily rolled back. You can create multiple mutable objects from one
+immutable reference to the fork. Moreover, there may be different forks of
+the same database snapshot.
+
+Forks are used during transaction and block processing. A fork
+[is successively passed](transactions.md#execute)
+to each transaction in the block to accumulate changes produced by the
+transactions, in a [patch](#patches). If one of the transactions in the block
```
Comment (slowli) [inline src/architecture/merkledb.md]: Nit: I'd say that the changes are accumulated in a fork (not a patch), i.e., remove "...in a patch." A patch is a result of _finalizing_ a fork.

File: src/architecture/merkledb.md [hunk 36]
```diff
@@ -0,0 +1,348 @@
...
+quits with an unhandled exception (i.e., raises `panic`) during execution, its
```
Comment (slowli) [inline src/architecture/merkledb.md]: Nit: a transaction is rolled back if it returns an error (a checked exception in Java terms), too, so I'd suggest to remove "(i.e., raises `panic`)".

File: misc/cspell.json [hunk 37]
```diff
@@ -49,6 +50,7 @@
     "markdownlint",
     "mempool",
     "Merkelized",
+	   "merklized",
```
Comment (slowli) [inline misc/cspell.json]: Is this spelling used in text? I thought we use "Merkelized" (which is mentioned directly above 🙃).

File: src/architecture/merkledb.md (hunk 28 above)
Comment (Maria-Nosyk) [reply src/architecture/merkledb.md]: done

File: src/architecture/storage.md (hunk 29 above)
Comment (Maria-Nosyk) [reply src/architecture/storage.md]: done

File: misc/cspell.json (hunk 37 above)
Comment (Maria-Nosyk) [reply misc/cspell.json]: deleted

File: src/architecture/storage.md (hunk 30 above)
Comment (Maria-Nosyk) [reply src/architecture/storage.md]: done

File: src/architecture/storage.md (hunk 31 above)
Comment (Maria-Nosyk) [reply src/architecture/storage.md]: edited

File: src/architecture/merkledb.md (hunk 32 above)
Comment (Maria-Nosyk) [reply src/architecture/merkledb.md]: done

File: src/architecture/merkledb.md (hunk 35 above)
Comment (Maria-Nosyk) [reply src/architecture/merkledb.md]: done

File: src/architecture/merkledb.md (hunk 36 above)
Comment (Maria-Nosyk) [reply src/architecture/merkledb.md]: done

File: src/architecture/merkledb.md (hunk 27 above)
Comment (Maria-Nosyk) [reply src/architecture/merkledb.md]: corrected

File: src/architecture/merkledb.md (hunk 34 above)
Comment (Maria-Nosyk) [reply src/architecture/merkledb.md]: deleted mentioning of Protobuf serialization

File: src/architecture/merkledb.md (hunk 33 above)
Comment (Maria-Nosyk) [reply src/architecture/merkledb.md]: deleted info about indices here and in the storage.md
>Indices are implicitly defined by the items order.

moved to the "Implementation Details" note of the `ListIndex` description

**Pull request merged**